src/frontend/src/optimizer/rule/apply_scan.rs

```diff
@@ -54,8 +55,9 @@ impl Rule for ApplyScanRule {
             // See the fourth section of Unnesting Arbitrary Queries for how to do the optimization.
             let mut exprs: Vec<ExprImpl> = correlated_indices
                 .into_iter()
-                .map(|correlated_index| {
-                    let (col_index, data_type) = column_mapping.get(&correlated_index).unwrap();
+                .enumerate()
+                .map(|(i, _)| {
```

Comment:
Looks like a nontrivial change. Could you elaborate why we should use enumerate indices rather than indices from `correlated_indices`?

cc @likg227

Reply:
For `Apply` operator its correlated_id refers to some correlated_input_ref which contain field `index` used to recored which column it point to.
After the planner and before optimizer, `index` of `correlated_input_ref` means it need to refer the left input of its corresponding `Apply` exactly with the `index`th column.
However, after TranslateApplyRule its we change the left input of Apply but without changing its corresponding `index` of `correlated_input_ref`. Luckily, we can use correlated_indices to help `correlated_input_ref` definitely find what column its refer to.
So `correlated_indices` of `Apply`, its meaning change after TranslateApplyRule, but we should no keep changing `correlated_indices` by any other rule because of the left input of `Apply` is always the same (called D).

Reply:
So it's actually a bug fix of previous implementation. Do I understand this correctly?

Reply:
Yes.

Reply:
Cool! Thanks for sharing the details.